pkg/snapstore/gcs_snapstore.go

```
@@ -60,21 +60,18 @@ const (
 // NewGCSSnapStore create new GCSSnapStore from shared configuration with specified bucket.
 func NewGCSSnapStore(config *brtypes.SnapstoreConfig) (*GCSSnapStore, error) {
 	ctx := context.TODO()
-	var emulatorConfig gcsEmulatorConfig
-	emulatorConfig.enabled = isEmulatorEnabled()
-	var opts []option.ClientOption // no need to explicitly set store credentials here since the Google SDK picks it up from the standard environment variable
 
-	if gcsApplicationCredentialsPath, isSet := os.LookupEnv(getEnvPrefixString(config.IsSource) + envStoreCredentials); isSet {
-		storageAPIEndpointFilePath := path.Join(path.Dir(gcsApplicationCredentialsPath), storageAPIEndpointFileName)
-		endpoint, err := getGCSStorageAPIEndpoint(storageAPIEndpointFilePath)
-		if err != nil {
-			return nil, fmt.Errorf("error getting storage API endpoint from %v", storageAPIEndpointFilePath)
-		}
-		if endpoint != "" {
-			opts = append(opts, option.WithEndpoint(endpoint))
-			if emulatorConfig.enabled {
-				emulatorConfig.endpoint = endpoint
-			}
+	var opts []option.ClientOption // no need to explicitly set store credentials here since the Google SDK picks it up from the standard environment variable
+	var emulatorConfig gcsEmulatorConfig
+	emulatorConfig.enabled = config.IsEmulatorEnabled || isEmulatorEnabled(config)
+	endpoint, err := getGCSStorageAPIEndpoint(config)
```

Just a thought: Can we add a validation here that if emulator is enabled but endpoint is not present or is empty, then return an error? If agreed, we can do the same for Azure as well.

---

Sure, makes sense to add this validation.

---

If the host is not present, we would get an error complaining that no host is present with a TCP connection failure which is quite self explanatory, in my opinion. I don't think this extra check is necessary.

With an empty "domain" file for ABS:

```
FATA[0015] failed to create snapstore from configured storage provider: failed to get properties of the container i585850-test with error: Get "https://devstoreaccount1./xxxxxx-test?restype=container": dial tcp: lookup devstoreaccount1.: no such host  actor=backup-restore-server
```

---

It would be better to throw the error early, than to try and decipher it from the URL in the error message.

---

@renormalize Btw, in ABS snapstore, we can never end up with URL `https://devstoreaccount1./xxxxxx-test?restype=container`, since domain is first defaulted to `brtypes.AzureBlobStorageGlobalDomain (blob.core.windows.net)`. So even if no domain is provided by user, the URL should still look like `https://devstoreaccount1.blob.core.windows.net/xxxxxx-test?restype=container`. And if emulator is enabled but domain isn't provided, the URL would look like `http://blob.core.windows.net/devstoreaccount1/xxxxxx-test?restype=container`. The `domain` will never be empty. It makes sense to throw the error early in the code, rather than try and figure out the malformed URL from the error logs.